packages/synapse-react-client/src/components/DataGrid/DataGrid.tsx

```diff
@@ -0,0 +1,695 @@
+      if (webSocketRef.current) {
+        webSocketRef.current.close()
+        webSocketRef.current = null
+      }
+    }
+  }, [sessionId, replicaId])
+
+  // Grid rows and columns
+  type DataGridRow = { [key: string]: string | number }
+  const [gridReady, setGridReady] = useState<boolean>(false)
+  const [rowValues, setRowValues] = useState<DataGridRow[]>(
+    modelRowsToGrid(modelSnapshotRef.current),
+  )
+  const [colValues, setColValues] = useState<Column[]>(
+    modelColsToGrid(modelSnapshotRef.current),
+  )
+  const [prevRows, setPrevRows] = useState(rowValues)
+
+  // Query Input can either be an empty string, a SQL query, or a session ID
+  const parseQueryInput = (input: string) => {
+    const trimmedInput = input.trim()
+    if (!trimmedInput) {
+      return { type: 'empty', input: '' }
+    } else if (trimmedInput.toUpperCase().startsWith('SELECT')) {
+      return { type: 'sql', input: trimmedInput }
+    } else if (/^N\w*=$/.test(trimmedInput)) {
+      return { type: 'sessionId', input: trimmedInput }
+    }
+    return { type: 'unknown', input: trimmedInput }
+  }
+
+  const createReplicaId = async (
+    synapseClient: SynapseClient,
+    sessionId: string,
+  ) => {
+    try {
+      console.log('Creating replica for session ID:', sessionId)
+      const replica =
+        await synapseClient.gridServicesClient.postRepoV1GridSessionIdReplica({
+          sessionId,
+          createReplicaRequest: {
+            gridSessionId: sessionId,
+          },
+        } as PostRepoV1GridSessionIdReplicaRequest)
+      return replica
+    } catch (error) {
+      console.error('Failed to create replica:', error)
+      return
+    }
+  }
+
+  // Convert model rows to a format suitable for DataSheetGrid
+  function modelRowsToGrid(modelSnapshot: ModelSnapshot): DataGridRow[] {
+    if (!modelSnapshot) return []
+    const { columnNames, columnOrder, rows } = modelSnapshot
+    const gridRows = rows.map(row => {
+      const rowObj: { [key: string]: any } = {}
+      // Use columnOrder to determine which columnNames to use and in what order
+      columnOrder.forEach((index: number) => {
+        const columnName = columnNames[index]
+        if (columnName) {
+          rowObj[columnName] = row[index]
+        }
+      })
+      return rowObj
+    })
+    return gridRows
+  }
+
+  // Convert model columns to a format suitable for DataSheetGrid
+  function modelColsToGrid(modelSnapshot: ModelSnapshot): Column[] {
+    if (!modelSnapshot) return []
+    const { columnNames, columnOrder } = modelSnapshot
+    const gridCols: Column[] = columnOrder.map((index: number) => {
+      return {
+        ...keyColumn(
+          columnNames[index],
+          createTextColumn({ continuousUpdates: false }),
+        ),
+        title: columnNames[index],
+      }
+    })
+    return gridCols
+  }
+
+  // Function to apply grid changes to a model
+  function gridToModel(gridRows: DataGridRow[], model: Model): Model {
```

Comment:
Model updates must use the API so that the clock is updated and patches can be computed.